python/source/state_representation/bind_state.cpp

```diff
@@ -0,0 +1,66 @@
+      .value("STATE", StateType::STATE)
+      .value("CARTESIANSTATE", StateType::CARTESIANSTATE)
+      .value("DUALQUATERNIONSTATE", StateType::DUALQUATERNIONSTATE)
+      .value("JOINTSTATE", StateType::JOINTSTATE)
+      .value("JACOBIANMATRIX", StateType::JACOBIANMATRIX)
+      .value("GEOMETRY_SHAPE", StateType::GEOMETRY_SHAPE)
+      .value("GEOMETRY_ELLIPSOID", StateType::GEOMETRY_ELLIPSOID)
+      .value("PARAMETER_DOUBLE", StateType::PARAMETER_DOUBLE)
+      .value("PARAMETER_DOUBLE_ARRAY", StateType::PARAMETER_DOUBLE_ARRAY)
+      .value("PARAMETER_BOOL", StateType::PARAMETER_BOOL)
+      .value("PARAMETER_BOOL_ARRAY", StateType::PARAMETER_BOOL_ARRAY)
+      .value("PARAMETER_STRING", StateType::PARAMETER_STRING)
+      .value("PARAMETER_STRING_ARRAY", StateType::PARAMETER_STRING_ARRAY)
+      .value("PARAMETER_CARTESIANSTATE", StateType::PARAMETER_CARTESIANSTATE)
+      .value("PARAMETER_CARTESIANPOSE", StateType::PARAMETER_CARTESIANPOSE)
+      .value("PARAMETER_JOINTSTATE", StateType::PARAMETER_JOINTSTATE)
+      .value("PARAMETER_JOINTPOSITIONS", StateType::PARAMETER_JOINTPOSITIONS)
+      .value("PARAMETER_ELLIPSOID", StateType::PARAMETER_ELLIPSOID)
+      .value("PARAMETER_MATRIX", StateType::PARAMETER_MATRIX)
+      .value("PARAMETER_VECTOR", StateType::PARAMETER_VECTOR)
+      .export_values();
+}
+
+void state(py::module_& m) {
+  py::class_<State> c(m, "State");
+
+  m.def("swap", py::overload_cast<State&, State&>(&state_representation::swap), "Swap the values of the two States", "state1"_a, "state2"_a);
+
+  c.def(py::init(), "Empty constructor");
+  c.def(py::init<const StateType&>(), "Constructor only specifying the type of the state from the StateType enumeration", "type"_a);
+  c.def(py::init<const StateType&, const std::string&, const bool&>(), "Constructor with name specification", "type"_a, "name"_a, "empty"_a=true);
+  c.def(py::init<const State&>(), "Copy constructor from another State", "state"_a);
+
+  c.def("get_type", &State::get_type, "Getter of the type attribute");
+  c.def("is_empty", &State::is_empty, "Getter of the empty attribute");
+  c.def("set_empty", &State::set_empty, "Setter of the empty attribute to true");
+  c.def("set_filled", &State::set_filled, "Setter of the empty attribute to false and also reset the timestamp");
+  c.def("get_timestamp", &State::get_timestamp, "Getter of the timestamp attribute");
+  c.def("reset_timestamp", &State::reset_timestamp, "Reset the timestamp attribute to now");
+  c.def("get_name", &State::get_name, "Getter of the name");
+  c.def("set_name", &State::set_name, "Setter of the name");
+
+  c.def("is_deprecated", py::overload_cast<const std::chrono::seconds&>(&State::is_deprecated<std::ratio<1>>), "Check if the state is deprecated given a certain time delay in seconds");
+  c.def("is_deprecated", py::overload_cast<const std::chrono::milliseconds&>(&State::is_deprecated<std::milli>), "Check if the state is deprecated given a certain time delay in milliseconds");
```

Comment:
How does that work? You would call `is_deprecated(10)` which version of the function would be called as python arguments are not typed? Also in c++, the problem with those `chrono` is that they only accept integers. So if we have a control loop of `> 1000Hz` (i.e. `< 1ms`) we need also the version that accepts `nanoseconds`

Reply:
Ah, I had two different things going on here that I forgot to clean up.

The first is that templated binding does not work for python, instead you have to overload specific templated instances (here for example passing the ratio template value to the `is_deprecated` function). I was testing this out so I made two variants.

The second is that `chrono` types are automatically converted to python `datetime` types. In particular, a `chrono::duration` becomes a `datetime.timedelta`, and uses a precision of `microseconds`. From the docs:

> any precision in the duration greater than microseconds is lost

Ultimately the best way forward for this function is not to do any overloading, but to bind a variant of the function with the maximum supported resolution of microseconds.
```c++
c.def("is_deprecated", &State::is_deprecated<std::micro>, "...");
```

Reply:
> When we have two function in c++ where the signature differs only by the type of input variables (same number arguments and same name). Is it able to resolve it?

Python does not support function overloading, but pybind11 takes care of it for us. By using the `py::overload_cast` macro, it creates a wrapper method under the given name (if it does not already exist) and adds the overloaded signature to a list of methods in the local namespace. The interpreter will then try to match the arguments given to the high-level named function to each signature in the list. They are evaluated in the order they are defined - whichever signature matches the arguments first is used.

I'm pretty sure that an overload of `foo(int)` and `foo(double)` wouldn't make much sense, because in the python `foo(2.0)` would be implicitly 

See more info in this section of the docs:
https://pybind11.readthedocs.io/en/stable/classes.html#overloaded-methods


I made a quick test case to demonstrate. Say we have two functions:

```c++
int foo(int a) {
  return a;
}

int foo(double a) {
  return static_cast<int>(2 * a);
}

PYBIND11_MODULE(foo_example, m) {
  m.def("foo", py::overload_cast<int>(&foo), "take an int, return it");
  m.def("foo", py::overload_cast<double>(&foo), "take a double, return double");
}
```

If we then check the python representation:
```python
from foo_example import foo

help(foo)
```

```text
foo(...) method of builtins.PyCapsule instance
    foo(*args, **kwargs)
    Overloaded function.
    
    1. foo(arg0: int) -> int
    
    take an int, return it
    
    2. foo(arg0: float) -> int
    
    take a double, return double
```

Even though the types (int and double) are implicitly convertible in Python, it still gives operator preference to the matching type. For ambiguous casting, the order of definition would be the tie-breaker. 
```python
foo(1)
>> 1

foo(1.0)
>> 2
```

Reply:
Thanks a lot for the explanation. This is very convenient